File: app/src/test/java/com/simplemobiletools/calculator/MainActivityTest.java
package com.simplemobiletools.calculator;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.annotation.Config;

import butterknife.ButterKnife;

import static junit.framework.Assert.assertEquals;

@RunWith(MyTestRunner.class)
@Config(constants = BuildConfig.class, manifest = "app/src/main/AndroidManifest.xml", sdk = 21)
public class MainActivityTest {
    MainActivity activity;

    @Before
    public void setUp() {
        activity = Robolectric.setupActivity(MainActivity.class);
        ButterKnife.bind(activity);
    }

    @Test
    public void addSimpleDigit() {
        activity.getCalc().addDigit(2);
        assertEquals("2", getDisplayedNumber());
    }

    @Test
    public void removeLeadingZero() {
        activity.getCalc().addDigit(0);
        activity.getCalc().addDigit(5);
        assertEquals("5", getDisplayedNumber());
    }

    @Test
    public void additionTest() {
        String res = calcResult(-1.2, Constants.PLUS, 3.4);
        assertEquals("2.2", res);
        checkFormula("-1.2+3.4");
    }

    @Test
    public void subtractionTest() {
        String res = calcResult(7.8, Constants.MINUS, 2.5);
        assertEquals("5.3", res);
        checkFormula("7.8-2.5");
    }

    @Test
    public void multiplyTest() {
        String res = calcResult(-3.2, Constants.MULTIPLY, 6.6);
        assertEquals("-21.12", res);
        checkFormula("-3.2*6.6");
    }

    @Test
    public void divisionTest() {
        String res = calcResult(18.25, Constants.DIVIDE, 5);
        assertEquals("3.65", res);
        checkFormula("18.25/5");
    }

    @Test
    public void divisionByZero_returnsZero() {
        String res = calcResult(6, Constants.DIVIDE, 0);
        assertEquals("0", res);
        checkFormula("6/0");
    }

    @Test
    public void moduloTest() {
        String res = calcResult(6.5, Constants.MODULO, 3);
        assertEquals("0.5", res);
        checkFormula("6.5%3");
    }

    @Test
    public void powerTest() {
        String res = calcResult(3, Constants.POWER, 6);
        assertEquals("729", res);
        checkFormula("3^6");
    }

    @Test
    public void rootTest() {
        setDouble(16);
        handleOperation(Constants.ROOT);
        assertEquals("4", getDisplayedNumber());
        checkFormula("√16");
    }

    @Test
    public void clearBtnSimpleTest() {
        setDouble(156);
        activity.getCalc().handleClear();
        assertEquals("15", getDisplayedNumber());
    }

    @Test
    public void clearBtnComplexTest() {
        setDouble(-26);
        activity.getCalc().handleClear();
        assertEquals("-2", getDisplayedNumber());
        activity.getCalc().handleClear();
        assertEquals("0", getDisplayedNumber());
    }

    @Test
    public void clearBtnLongClick_resetsEverything() {
        calcResult(-1.2, Constants.PLUS, 3.4);
        activity.getCalc().handleReset();
        handleOperation(Constants.PLUS);
        setDouble(3);
        activity.getCalc().handleResult();
        assertEquals("3", getDisplayedNumber());
        checkFormula("");
    }

    @Test
    public void complexTest() {
        setDouble(-12.2);
        handleOperation(Constants.PLUS);
        setDouble(21);
        handleOperation(Constants.MINUS);
        assertEquals("8.8", getDisplayedNumber());
        checkFormula("-12.2+21");

        setDouble(1.6);
        activity.getCalc().handleEquals();
        assertEquals("7.2", getDisplayedNumber());
        checkFormula("8.8-1.6");
        activity.getCalc().handleEquals();
        assertEquals("5.6", getDisplayedNumber());
        checkFormula("7.2-1.6");

        handleOperation(Constants.MULTIPLY);
        setDouble(5);
        handleOperation(Constants.DIVIDE);
        assertEquals("28", getDisplayedNumber());
        checkFormula("5.6*5");

        setDouble(4);
        handleOperation(Constants.MODULO);
        assertEquals("7", getDisplayedNumber());
        checkFormula("28/4");

        setDouble(5);
        handleOperation(Constants.POWER);
        assertEquals("2", getDisplayedNumber());
        checkFormula("7%5");

        setDouble(8);
        handleOperation(Constants.ROOT);
        assertEquals("16", getDisplayedNumber());
        checkFormula("√256");

        activity.getCalc().handleClear();
        assertEquals("1", getDisplayedNumber());
    }

    private void setDouble(double d) {
        activity.setValueDouble(d);
    }

    private void handleOperation(String operation) {
        activity.getCalc().handleOperation(operation);
    }

    private void checkFormula(String desired) {
        assertEquals(desired, getDisplayedFormula());
    }

    private String calcResult(double baseValue, String operation, double secondValue) {
        setDouble(baseValue);
        handleOperation(operation);
        setDouble(secondValue);
        activity.getCalc().handleResult();
        return getDisplayedNumber();
    }

    private String getDisplayedNumber() {
        return activity.getCalc().getDisplayedNumber();
    }

    private String getDisplayedFormula() {
        return activity.getCalc().getDisplayedFormula();
    }
}

File: app/src/test/java/com/simplemobiletools/calculator/MainActivityTest.kt
package com.simplemobiletools.calculator

import com.simplemobiletools.calculator.activities.MainActivity
import com.simplemobiletools.calculator.helpers.*
import com.simplemobiletools.calculator.BuildConfig
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.math.BigDecimal

@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class, sdk = [21])
class MainActivityTest {
    private lateinit var activity: MainActivity

    private fun getDisplayedNumber() = activity.calc.displayedNumber

    @Before
    fun setUp() {
        activity = Robolectric.setupActivity(MainActivity::class.java)
    }

    @Test
    fun addSimpleDigit() {
        activity.calc.addDigit(2)
        assertEquals("2", getDisplayedNumber())
    }

    @Test
    fun removeLeadingZero() {
        activity.calc.addDigit(0)
        activity.calc.addDigit(5)
        assertEquals("5", getDisplayedNumber())
    }

    @Test
    fun additionTest() {
        val res = calcResult(BigDecimal.valueOf(-1.2), PLUS, BigDecimal.valueOf(3.4))
        assertEquals("2.2", res)
        checkFormula("-1.2+3.4")
    }

    @Test
    fun subtractionTest() {
        val res = calcResult(BigDecimal.valueOf(7.8), MINUS, BigDecimal.valueOf(2.5))
        assertEquals("5.3", res)
        checkFormula("7.8-2.5")
    }

    @Test
    fun multiplyTest() {
        val res = calcResult(BigDecimal.valueOf(-3.2), MULTIPLY, BigDecimal.valueOf(6.6))
        assertEquals("-21.12", res)
        checkFormula("-3.2*6.6")
    }

    @Test
    fun divisionTest() {
        val res = calcResult(BigDecimal.valueOf(18.25), DIVIDE, BigDecimal.valueOf(5.0))
        assertEquals("3.65", res)
        checkFormula("18.25/5")
    }

    @Test
    fun divisionByZero_returnsZero() {
        val res = calcResult(BigDecimal.valueOf(6.0), DIVIDE, BigDecimal.valueOf(0.0))
        assertEquals("0", res)
        checkFormula("6/0")
    }

    /*  @Test
      fun moduloTest() {
          val res = calcResult(6.5, MODULO, 3.0)
          assertEquals("0.5", res)
          checkFormula("6.5%3")
      }*/

    @Test
    fun powerTest() {
        val res = calcResult(BigDecimal.valueOf(3.0), POWER, BigDecimal.valueOf(6.0))
        assertEquals("729", res)
        checkFormula("3^6")
    }

    @Test
    fun rootTest() {
        setBigDecimal(BigDecimal(16.0))
        handleOperation(ROOT)
        assertEquals("4", getDisplayedNumber())
        checkFormula("√16")
    }

    @Test
    fun clearBtnSimpleTest() {
        setBigDecimal(BigDecimal(156.0))
        activity.calc.handleClear()
        assertEquals("15", getDisplayedNumber())
    }

    @Test
    fun clearBtnComplexTest() {
        setBigDecimal(BigDecimal(-26.0))
        activity.calc.handleClear()
        assertEquals("-2", getDisplayedNumber())
        activity.calc.handleClear()
        assertEquals("0", getDisplayedNumber())
    }

    @Test
    fun clearBtnLongClick_resetsEverything() {
        calcResult(BigDecimal.valueOf(-1.2), PLUS, BigDecimal.valueOf(3.4))
        activity.calc.handleReset()
        handleOperation(PLUS)
        setBigDecimal(BigDecimal.valueOf(3.0))
        activity.calc.handleResult()
        assertEquals("3", getDisplayedNumber())
        checkFormula("")
    }

    @Test
    fun complexTest() {
        setBigDecimal(BigDecimal.valueOf(-12.2))
        handleOperation(PLUS)
        setBigDecimal(BigDecimal(21.0))
        handleOperation(MINUS)
        assertEquals("8.8", getDisplayedNumber())
        checkFormula("-12.2+21")

        setBigDecimal(BigDecimal(1.6))
        activity.calc.handleEquals()
        assertEquals("7.2", getDisplayedNumber())
        checkFormula("8.8-1.6")
        activity.calc.handleEquals()
        assertEquals("5.6", getDisplayedNumber())
        checkFormula("7.2-1.6")

        handleOperation(MULTIPLY)
        setBigDecimal(BigDecimal(5.0))
        handleOperation(DIVIDE)
        assertEquals("28", getDisplayedNumber())
        checkFormula("5.6*5")

        setBigDecimal(BigDecimal(4.0))
        handleOperation(DIVIDE)
        assertEquals("7", getDisplayedNumber())
        checkFormula("28/4")

        setBigDecimal(BigDecimal(5.0))
        handleOperation(POWER)
        assertEquals("2", getDisplayedNumber())
        checkFormula("7%5")

        setBigDecimal(BigDecimal(8.0))
        handleOperation(ROOT)
        assertEquals("16", getDisplayedNumber())
        checkFormula("√256")

        activity.calc.handleClear()
        assertEquals("1", getDisplayedNumber())
    }

    private fun setBigDecimal(d: BigDecimal) {
        activity.setValueBigDecimal(d)
    }

    private fun handleOperation(operation: String) {
        activity.calc.handleOperation(operation)
    }

    private fun checkFormula(desired: String) {
        assertEquals(desired, activity.calc.displayedFormula)
    }

    private fun calcResult(baseValue: BigDecimal, operation: String, secondValue: BigDecimal): String? {
        setBigDecimal(baseValue)
        handleOperation(operation)
        setBigDecimal(secondValue)
        activity.calc.handleResult()
        return getDisplayedNumber()
    }
}
